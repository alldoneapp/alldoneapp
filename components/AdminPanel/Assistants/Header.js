import React from 'react'
import { View, Text, StyleSheet } from 'react-native'

import { translate } from '../../../i18n/TranslationService'
import styles, { colors } from '../../styles/global'

export default function Header({ assistantsAmount }) {
    const parseText = number => {
        if (number === 0) {
            return translate('No assistants yet')
        } else if (number > 1) {
            return number + ` ${translate('Assistants')}`
        }
        return number + ` ${translate('Assistant')}`
    }

    return (
        <View style={localStyles.header}>
            <Text style={[styles.title6, { color: colors.Text01 }]}>{translate('AI Assistants')}</Text>
            <View style={localStyles.headerCaption}>
                <Text style={[styles.caption2, { color: colors.Text02 }]}>{parseText(assistantsAmount)}</Text>
            </View>
        </View>
    )
}

const localStyles = StyleSheet.create({
    header: {
        paddingTop: 32,
        paddingBottom: 12,
        alignItems: 'flex-end',
        flexDirection: 'row',
    },
    headerCaption: {
        marginLeft: 16,
        height: 22,
        justifyContent: 'center',
    },
})
