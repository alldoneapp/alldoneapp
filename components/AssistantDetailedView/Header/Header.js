import React, { useState, useEffect } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { useSelector } from 'react-redux'

import TitlePresentation from './TitlePresentation'
import TitleEdition from './TitleEdition'
import LastEdition from './LastEdition'
import BackButton from './BackButton'
import CopyLinkButton from '../../UIControls/CopyLinkButton'
import DVHamburgButton from '../../UIControls/DVHamburgButton'
import OpenInNewWindowButton from '../../UIControls/OpenInNewWindowButton'
import styles, { colors } from '../../styles/global'
import { DV_TAB_ASSISTANT_CHAT, DV_TAB_ASSISTANT_NOTE } from '../../../utils/TabNavigationConstants'
import DvBotButton from '../../UIControls/DvBotButton'
import BotLine from '../../ChatsView/ChatDV/BotLine/BotLine'

export default function Header({
    projectId,
    projectDetailedId,
    assistant,
    navigation,
    isGlobalAsisstant,
    isFullscreen,
    setFullscreen,
}) {
    const showGlobalSearchPopup = useSelector(state => state.showGlobalSearchPopup)
    const isMiddleScreen = useSelector(state => state.isMiddleScreen)
    const mobile = useSelector(state => state.smallScreenNavigation)
    const selectedTab = useSelector(state => state.selectedNavItem)
    const loggedUser = useSelector(state => state.loggedUser)
    const [editionMode, setEditionMode] = useState(false)
    const [showEllipsis, setShowEllipsis] = useState(false)

    const maxHeight =
        (selectedTab === DV_TAB_ASSISTANT_NOTE || selectedTab === DV_TAB_ASSISTANT_CHAT) && !editionMode ? 64 : 800

    const openTitleEdition = () => {
        setEditionMode(true)
    }

    const closeTitleEdition = () => {
        setEditionMode(false)
    }

    useEffect(() => {
        if (showGlobalSearchPopup && editionMode) {
            closeTitleEdition()
        }
    }, [showGlobalSearchPopup])

    const onTitleLayoutChange = ({ nativeEvent }) => {
        const { layout } = nativeEvent

        if (layout.height > maxHeight && !showEllipsis) {
            setShowEllipsis(true)
        } else if (layout.height <= maxHeight && showEllipsis) {
            setShowEllipsis(false)
        }
    }

    return (
        <View style={[localStyles.container, isFullscreen && { paddingBottom: 8 }]}>
            <View style={localStyles.upperHeader}>
                {isMiddleScreen && <BackButton assistant={assistant} />}

                {mobile && loggedUser.isAnonymous && (
                    <View style={localStyles.backButtonMobile}>
                        <DVHamburgButton navigation={navigation} />
                    </View>
                )}

                <View style={[localStyles.titleContainer, { maxHeight: maxHeight }]}>
                    {editionMode ? (
                        <TitleEdition
                            projectId={projectId}
                            assistant={assistant}
                            closeTitleEdition={closeTitleEdition}
                        />
                    ) : (
                        <View onLayout={onTitleLayoutChange}>
                            <TitlePresentation
                                openTitleEdition={openTitleEdition}
                                assistant={assistant}
                                disabled={isGlobalAsisstant || loggedUser.isAnonymous}
                            />
                        </View>
                    )}
                    {showEllipsis && !editionMode && (
                        <Text style={[localStyles.ellipsis, { right: mobile ? 32 : 80 }]}>...</Text>
                    )}
                </View>
            </View>
            {!isFullscreen && (
                <View style={localStyles.bottomHeader}>
                    <View style={{ flex: 1 }} />
                    <LastEdition assistant={assistant} />
                    <CopyLinkButton style={{ marginRight: 8 }} />
                    <DvBotButton
                        navItem={DV_TAB_ASSISTANT_CHAT}
                        projectId={projectDetailedId}
                        assistantId={assistant.uid}
                    />
                    <OpenInNewWindowButton />
                </View>
            )}
            {isFullscreen && selectedTab === DV_TAB_ASSISTANT_CHAT && (
                <View style={localStyles.bottomHeader}>
                    <BotLine
                        setFullscreen={setFullscreen}
                        objectId={assistant.uid}
                        assistantId={assistant.uid}
                        projectId={projectDetailedId}
                        objectType={'assistants'}
                    />
                </View>
            )}
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        paddingBottom: 24,
    },
    upperHeader: {
        flexDirection: 'row',
        justifyContent: 'space-between',
    },
    titleContainer: {
        marginRight: 'auto',
        flex: 1,
        maxHeight: 800,
        overflowY: 'hidden',
    },
    bottomHeader: {
        marginTop: 22,
        flexDirection: 'row',
        alignItems: 'center',
    },
    backButtonMobile: {
        left: -16,
    },
    ellipsis: {
        ...styles.title4,
        color: colors.Text01,
        backgroundColor: '#ffffff',
        paddingHorizontal: 8,
        position: 'absolute',
        bottom: 0,
        right: 0,
    },
})
