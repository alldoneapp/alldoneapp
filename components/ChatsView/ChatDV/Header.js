import React, { useState } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import BackButton from './BackButton'
import TagList from './TagList'
import BotLine from './BotLine/BotLine'
import ChatTitle from './ChatTitle'
import ChatTitleEdition from './ChatTitleEdition'
import { useSelector } from 'react-redux'
import { DV_TAB_CHAT_BOARD, DV_TAB_CHAT_NOTE } from '../../../utils/TabNavigationConstants'
import styles, { colors } from '../../styles/global'
import SharedHelper from '../../../utils/SharedHelper'

const Header = ({ projectId, chat, isFullscreen, setFullscreen }) => {
    const mobile = useSelector(state => state.smallScreenNavigation)
    const isMiddleScreen = useSelector(state => state.isMiddleScreen)
    const selectedTab = useSelector(state => state.selectedNavItem)
    const loggedUser = useSelector(state => state.loggedUser)
    const [editionMode, setEditionMode] = useState(false)
    const [title, setTitle] = useState(chat.title)
    const [showEllipsis, setShowEllipsis] = useState(false)
    const maxHeight = (selectedTab === DV_TAB_CHAT_BOARD || selectedTab === DV_TAB_CHAT_NOTE) && !editionMode ? 64 : 350

    const onTitleLayoutChange = ({ nativeEvent }) => {
        const { layout } = nativeEvent

        if (layout.height > maxHeight && !showEllipsis) {
            setShowEllipsis(true)
        } else if (layout.height <= maxHeight && showEllipsis) {
            setShowEllipsis(false)
        }
    }

    const accessGranted = SharedHelper.accessGranted(loggedUser, projectId)

    return (
        <>
            <View style={[localStyles.upperHeader, { marginLeft: mobile ? -16 : isMiddleScreen ? -56 : -72 }]}>
                {accessGranted && <BackButton isFullscreen={isFullscreen} />}
                {(!isFullscreen || selectedTab === DV_TAB_CHAT_BOARD) && (
                    <View style={[localStyles.titleContainer, { maxHeight: maxHeight }]}>
                        {editionMode ? (
                            <ChatTitleEdition
                                projectId={projectId}
                                title={title}
                                setTitle={setTitle}
                                chat={chat}
                                closeTitleEdition={() => setEditionMode(false)}
                            />
                        ) : (
                            <View onLayout={onTitleLayoutChange}>
                                <ChatTitle
                                    projectId={projectId}
                                    title={title}
                                    openTitleEdition={() => setEditionMode(true)}
                                    disabled={!accessGranted}
                                />
                            </View>
                        )}
                        {showEllipsis && !editionMode && (
                            <Text style={[localStyles.ellipsis, { right: mobile ? 32 : 80 }]}>...</Text>
                        )}
                    </View>
                )}
            </View>
            {!isFullscreen && (
                <View style={localStyles.bottomHeader}>
                    <TagList projectId={projectId} chat={chat} />
                </View>
            )}
            {isFullscreen && selectedTab === DV_TAB_CHAT_BOARD && (
                <View style={localStyles.bottomHeader}>
                    <BotLine
                        setFullscreen={setFullscreen}
                        objectId={chat.id}
                        assistantId={chat.assistantId}
                        projectId={projectId}
                        objectType={'chats'}
                    />
                </View>
            )}
        </>
    )
}

const localStyles = StyleSheet.create({
    titleContainer: {
        flex: 1,
        marginLeft: 72,
        marginRight: 'auto',
        maxHeight: 350,
        overflowY: 'hidden',
    },
    upperHeader: {
        flexDirection: 'row',
    },
    bottomHeader: {
        paddingTop: 32,
        flexDirection: 'row',
    },
    ellipsis: {
        ...styles.title4,
        color: colors.Text01,
        backgroundColor: '#ffffff',
        paddingHorizontal: 8,
        position: 'absolute',
        bottom: 0,
        right: 0,
    },
})

export default Header
