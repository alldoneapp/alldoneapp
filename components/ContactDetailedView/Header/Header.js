import React, { useState, useRef, useEffect } from 'react'
import { StyleSheet, TouchableOpacity, View } from 'react-native'
import { useSelector } from 'react-redux'

import UserTitle from '../../UserDetailedView/Header/UserTitle'
import Indicator from './Indicator'
import BackButton from '../../UserDetailedView/Header/BackButton'
import MyPlatform from '../../MyPlatform'
import { colors } from '../../styles/global'
import Button from '../../UIControls/Button'
import CustomTextInput3 from '../../Feeds/CommentsTextInput/CustomTextInput3'
import TagList from './TagList'
import { exitsOpenModals } from '../../ModalsManager/modalsManager'
import { Dismissible } from 'react-dismissible'
import SharedHelper from '../../../utils/SharedHelper'
import { DV_TAB_CONTACT_CHAT } from '../../../utils/TabNavigationConstants'
import BotLine from '../../ChatsView/ChatDV/BotLine/BotLine'
import { setProjectContactName } from '../../../utils/backends/Contacts/contactsFirestore'

export default function Header({ contact, disabled, isFullscreen, setFullscreen, projectId }) {
    const loggedUser = useSelector(state => state.loggedUser)
    const project = useSelector(state => state.loggedUserProjectsMap[projectId])
    const isMiddleScreen = useSelector(state => state.isMiddleScreen)
    const selectedNavItem = useSelector(state => state.selectedNavItem)
    const showGlobalSearchPopup = useSelector(state => state.showGlobalSearchPopup)
    const [inputWidth, setInputWidth] = useState(0)
    const [focused, setFocused] = useState(true)
    const [editContactName, setEditContactName] = useState(false)
    const [contactName, setContactName] = useState(contact.displayName)

    const inputText = useRef(null)
    const mainContainer = useRef(null)
    const buttonsContainer = useRef(null)

    const onFocus = () => {
        setFocused(true)
    }

    const onBlur = () => {
        setFocused(false)
    }

    const updateInputWidth = async () => {
        if (mainContainer.current) {
            const promises = []
            promises.push(MyPlatform.getElementWidth(mainContainer.current))
            promises.push(MyPlatform.getElementWidth(buttonsContainer.current))
            const [containerWidth, buttonsContainerWidth] = await Promise.all(promises)
            setInputWidth(containerWidth - buttonsContainerWidth)
        }
    }

    const toggleEditContactName = () => {
        setEditContactName(state => !state)
    }

    const onChangeText = value => {
        setContactName(value.replaceAll('\n', ''))
    }

    const cancelEditForm = () => {
        setContactName(contact.displayName)
        toggleEditContactName()
    }

    const applyTitleChanges = () => {
        if (contactName.trim()) {
            setProjectContactName(projectId, contact, contact.uid, contactName.trim(), contact.displayName)
        } else {
            setContactName(contact.displayName)
        }
        toggleEditContactName()
    }

    const onInputKeyPress = key => {
        if (key === 'Enter') {
            applyTitleChanges()
        } else if (key === 'Escape') {
            cancelEditForm()
        }
    }

    useEffect(() => {
        if (showGlobalSearchPopup && editContactName) cancelEditForm()
    }, [showGlobalSearchPopup, editContactName])

    const accessGranted = SharedHelper.accessGranted(loggedUser, projectId)
    const disabledBtn = contact.displayName === contactName || contactName === ''

    return (
        <View style={localStyles.container}>
            <View style={[localStyles.upperHeader, isFullscreen && { paddingBottom: 16 }]}>
                {isMiddleScreen && accessGranted && (
                    <View style={localStyles.backButtonMobile}>
                        <BackButton user={contact} projectIndex={project.index} />
                    </View>
                )}

                {!editContactName ? (
                    <TouchableOpacity
                        style={[localStyles.upperHeader, { flex: 1 }]}
                        onPress={toggleEditContactName}
                        disabled={!accessGranted || disabled}
                    >
                        <View style={[localStyles.upperHeader, { flex: 1 }]}>
                            <View style={{ marginRight: 'auto', flex: 1 }}>
                                <UserTitle contact={contact} project={project} isContact={true} />
                            </View>
                            <View>
                                <Indicator />
                            </View>
                        </View>
                    </TouchableOpacity>
                ) : (
                    <View ref={mainContainer} style={[localStyles.inputContainer]} onLayout={updateInputWidth}>
                        <View
                            style={[
                                localStyles.textInputContainer,
                                focused ? localStyles.inputContainerFocused : localStyles.inputContainerBlurred,
                                { width: inputWidth },
                            ]}
                        >
                            <Dismissible
                                disabled={exitsOpenModals()}
                                click={true}
                                escape={true}
                                onDismiss={cancelEditForm}
                            >
                                <CustomTextInput3
                                    ref={inputText}
                                    numberOfLines={1}
                                    autoFocus={true}
                                    initialTextExtended={contactName}
                                    onKeyPress={onInputKeyPress}
                                    placeholderTextColor={colors.Text03}
                                    containerStyle={[
                                        localStyles.textInput,
                                        !focused ? localStyles.inputTextBlurred : undefined,
                                        { width: inputWidth },
                                    ]}
                                    onFocus={onFocus}
                                    onBlur={onBlur}
                                    onChangeText={onChangeText}
                                    projectId={projectId}
                                    projectIndex={project.index}
                                    placeholder="Write the name of the contact"
                                    singleLine={true}
                                    forceTriggerEnterActionForBreakLines={applyTitleChanges}
                                />
                            </Dismissible>
                        </View>

                        <View ref={buttonsContainer} style={localStyles.buttonsContainer}>
                            <Button
                                type={'secondary'}
                                icon={'x'}
                                buttonStyle={localStyles.secondaryBtn}
                                onPress={cancelEditForm}
                            />
                            <Button type={'primary'} icon={'save'} disabled={disabledBtn} onPress={applyTitleChanges} />
                        </View>
                    </View>
                )}
            </View>

            {!isFullscreen && (
                <View style={localStyles.bottomHeader}>
                    <TagList project={project} contact={contact} />
                </View>
            )}
            {isFullscreen && selectedNavItem === DV_TAB_CONTACT_CHAT && (
                <View style={localStyles.bottomHeader}>
                    <BotLine
                        setFullscreen={setFullscreen}
                        objectId={contact.uid}
                        assistantId={contact.assistantId}
                        projectId={projectId}
                        objectType={'contacts'}
                    />
                </View>
            )}
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        height: 140,
        flexDirection: 'column',
        justifyContent: 'space-between',
        paddingBottom: 24,
        overflow: 'hidden',
    },
    upperHeader: {
        flexDirection: 'row',
        justifyContent: 'space-between',
    },
    bottomHeader: {
        flexDirection: 'row',
        justifyItems: 'flex-end',
    },
    backButtonMobile: {
        left: -16,
    },
    inputContainer: {
        flex: 1,
        flexDirection: 'row',
        alignItems: 'center',
        marginTop: 28,
    },
    textInputContainer: {
        flex: 1,
        height: 40,
        borderWidth: 2,
        borderRadius: 4,
        overflow: 'hidden',
        flexDirection: 'row',
        alignItems: 'center',
    },
    inputContainerFocused: {
        borderWidth: 2,
        borderColor: colors.UtilityBlue200,
    },
    inputContainerBlurred: {
        borderWidth: 1,
        height: 42,
        borderColor: colors.Grey400,
    },
    inputTextBlurred: {
        height: 44,
        left: -1,
        color: colors.Text03,
    },
    textInput: {
        paddingHorizontal: 14,
        paddingTop: 0,
        height: 24,
    },
    buttonsContainer: {
        flexDirection: 'row',
    },
    secondaryBtn: {
        marginLeft: 8,
        marginRight: 8,
    },
})
