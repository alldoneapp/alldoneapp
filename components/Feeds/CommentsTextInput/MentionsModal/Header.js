import React from 'react'
import { StyleSheet, View } from 'react-native'

import HeaderTab from './HeaderTab'
import {
    MENTION_MODAL_CONTACTS_TAB,
    MENTION_MODAL_NOTES_TAB,
    MENTION_MODAL_TASKS_TAB,
    MENTION_MODAL_TOPICS_TAB,
    MENTION_MODAL_GOALS_TAB,
} from '../textInputHelper'

export default function Header({
    activeTab,
    changeTab,
    showHints,
    tasksAmount,
    goalsAmount,
    notesAmount,
    contactsAmount,
    chatsAmount,
}) {
    const activeTasksTab = () => {
        changeTab(MENTION_MODAL_TASKS_TAB)
    }
    const activeNotesTab = () => {
        changeTab(MENTION_MODAL_NOTES_TAB)
    }
    const activeGoalsTab = () => {
        changeTab(MENTION_MODAL_GOALS_TAB)
    }
    const activeContactsTab = () => {
        changeTab(MENTION_MODAL_CONTACTS_TAB)
    }
    const activeTopicsTab = () => {
        changeTab(MENTION_MODAL_TOPICS_TAB)
    }
    return (
        <View style={localStyles.container}>
            <HeaderTab
                text="Tasks"
                onPress={activeTasksTab}
                isActive={activeTab === MENTION_MODAL_TASKS_TAB}
                isNextShortcutTab={showHints && activeTab === MENTION_MODAL_TOPICS_TAB}
                badgeValue={tasksAmount}
            />
            <HeaderTab
                text="Goals"
                onPress={activeGoalsTab}
                isActive={activeTab === MENTION_MODAL_GOALS_TAB}
                isNextShortcutTab={showHints && activeTab === MENTION_MODAL_TASKS_TAB}
                badgeValue={goalsAmount}
            />
            <HeaderTab
                text="Notes"
                onPress={activeNotesTab}
                isActive={activeTab === MENTION_MODAL_NOTES_TAB}
                isNextShortcutTab={showHints && activeTab === MENTION_MODAL_GOALS_TAB}
                badgeValue={notesAmount}
            />
            <HeaderTab
                text="Contacts"
                onPress={activeContactsTab}
                isActive={activeTab === MENTION_MODAL_CONTACTS_TAB}
                isNextShortcutTab={showHints && activeTab === MENTION_MODAL_NOTES_TAB}
                badgeValue={contactsAmount}
            />
            <HeaderTab
                text="Chats"
                onPress={activeTopicsTab}
                isActive={activeTab === MENTION_MODAL_TOPICS_TAB}
                isNextShortcutTab={showHints && activeTab === MENTION_MODAL_CONTACTS_TAB}
                badgeValue={chatsAmount}
            />
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        marginHorizontal: 8,
        flexDirection: 'row',
        alignItems: 'center',
        marginBottom: 16,
    },
})
