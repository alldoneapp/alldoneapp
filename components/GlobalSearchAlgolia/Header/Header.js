import React, { useEffect } from 'react'
import { StyleSheet, View } from 'react-native'

import HeaderTab from './HeaderTab'
import {
    MENTION_MODAL_CONTACTS_TAB,
    MENTION_MODAL_GOALS_TAB,
    MENTION_MODAL_NOTES_TAB,
    MENTION_MODAL_TASKS_TAB,
    MENTION_MODAL_TOPICS_TAB,
} from '../../Feeds/CommentsTextInput/textInputHelper'

export default function Header({
    activeTab,
    setActiveTab,
    tasksResultAmount,
    goalsResultAmount,
    contactsResultAmount,
    notesResultAmount,
    chatsResultAmount,
    showShortcuts,
}) {
    useEffect(() => {
        document.addEventListener('keydown', onKeyDown)
        return () => document.removeEventListener('keydown', onKeyDown)
    })

    const changeTab = tab => {
        if (activeTab !== tab) {
            setActiveTab(tab)
        }
    }

    const onKeyDown = event => {
        const { key } = event
        if (key === 'Tab') {
            if (activeTab === MENTION_MODAL_TOPICS_TAB) {
                changeTab(MENTION_MODAL_TASKS_TAB)
            } else {
                changeTab(activeTab + 1)
            }
        }
    }

    const activeTasksTab = () => {
        changeTab(MENTION_MODAL_TASKS_TAB)
    }
    const activeNotesTab = () => {
        changeTab(MENTION_MODAL_NOTES_TAB)
    }
    const activeGoalsTab = () => {
        changeTab(MENTION_MODAL_GOALS_TAB)
    }
    const activeContactsTab = () => {
        changeTab(MENTION_MODAL_CONTACTS_TAB)
    }
    const activeTopicsTab = () => {
        changeTab(MENTION_MODAL_TOPICS_TAB)
    }

    return (
        <View style={localStyles.container}>
            <HeaderTab
                text={'Tasks'}
                onPress={activeTasksTab}
                isActive={activeTab === MENTION_MODAL_TASKS_TAB}
                isNextShortcutTab={activeTab === MENTION_MODAL_TOPICS_TAB}
                badgeValue={tasksResultAmount}
                showShortcuts={showShortcuts}
            />
            <HeaderTab
                text={'Goals'}
                onPress={activeGoalsTab}
                isActive={activeTab === MENTION_MODAL_GOALS_TAB}
                isNextShortcutTab={activeTab === MENTION_MODAL_TASKS_TAB}
                badgeValue={goalsResultAmount}
                showShortcuts={showShortcuts}
            />
            <HeaderTab
                text={'Notes'}
                onPress={activeNotesTab}
                isActive={activeTab === MENTION_MODAL_NOTES_TAB}
                isNextShortcutTab={activeTab === MENTION_MODAL_GOALS_TAB}
                badgeValue={notesResultAmount}
                showShortcuts={showShortcuts}
            />
            <HeaderTab
                text={'Contacts'}
                onPress={activeContactsTab}
                isActive={activeTab === MENTION_MODAL_CONTACTS_TAB}
                isNextShortcutTab={activeTab === MENTION_MODAL_NOTES_TAB}
                badgeValue={contactsResultAmount}
                showShortcuts={showShortcuts}
            />
            <HeaderTab
                text={'Chats'}
                onPress={activeTopicsTab}
                isActive={activeTab === MENTION_MODAL_TOPICS_TAB}
                isNextShortcutTab={activeTab === MENTION_MODAL_CONTACTS_TAB}
                badgeValue={chatsResultAmount}
                showShortcuts={showShortcuts}
            />
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        width: '100%',
        flexDirection: 'row',
        alignItems: 'center',
        marginTop: 24,
        marginBottom: 8,
        paddingHorizontal: 16,
    },
})
