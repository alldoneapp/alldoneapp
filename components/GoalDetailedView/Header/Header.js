import React, { useState, useEffect } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { useSelector } from 'react-redux'

import BackButton from './BackButton'
import TagList from './TagList'
import TitlePresentation from './TitlePresentation'
import TitleEdition from './TitleEdition'
import LastEdition from './LastEdition'
import CopyLinkButton from '../../UIControls/CopyLinkButton'
import DVHamburgButton from '../../UIControls/DVHamburgButton'
import OpenInNewWindowButton from '../../UIControls/OpenInNewWindowButton'
import { DV_TAB_GOAL_CHAT, DV_TAB_GOAL_NOTE } from '../../../utils/TabNavigationConstants'
import styles, { colors } from '../../styles/global'
import ProjectHelper from '../../SettingsView/ProjectsSettings/ProjectHelper'
import DvBotButton from '../../UIControls/DvBotButton'
import BotLine from '../../ChatsView/ChatDV/BotLine/BotLine'

export default function Header({ goal, projectId, navigation, accessGranted, isFullscreen, setFullscreen }) {
    const showGlobalSearchPopup = useSelector(state => state.showGlobalSearchPopup)
    const isMiddleScreen = useSelector(state => state.isMiddleScreen)
    const mobile = useSelector(state => state.smallScreenNavigation)
    const selectedTab = useSelector(state => state.selectedNavItem)
    const loggedUser = useSelector(state => state.loggedUser)
    const [editionMode, setEditionMode] = useState(false)
    const [showEllipsis, setShowEllipsis] = useState(false)
    const maxHeight = (selectedTab === DV_TAB_GOAL_CHAT || selectedTab === DV_TAB_GOAL_NOTE) && !editionMode ? 64 : 800

    const { completionMilestoneDate } = goal

    const openTitleEdition = () => {
        setEditionMode(true)
    }

    const closeTitleEdition = () => {
        setEditionMode(false)
    }

    useEffect(() => {
        if (showGlobalSearchPopup && editionMode) {
            closeTitleEdition()
        }
    }, [showGlobalSearchPopup])

    const onTitleLayoutChange = ({ nativeEvent }) => {
        const { layout } = nativeEvent

        if (layout.height > maxHeight && !showEllipsis) {
            setShowEllipsis(true)
        } else if (layout.height <= maxHeight && showEllipsis) {
            setShowEllipsis(false)
        }
    }

    const loggedUserIsGoalOwner = goal.ownerId === loggedUser.uid
    const loggedUserCanUpdateObject =
        loggedUserIsGoalOwner || !ProjectHelper.checkIfLoggedUserIsNormalUserInGuide(projectId)

    return (
        <View style={[localStyles.container, isFullscreen && { paddingBottom: 8 }]}>
            <View style={localStyles.upperHeader}>
                {isMiddleScreen && accessGranted && <BackButton projectId={projectId} goal={goal} />}

                {mobile && loggedUser.isAnonymous && (
                    <View style={localStyles.backButtonMobile}>
                        <DVHamburgButton navigation={navigation} />
                    </View>
                )}

                <View style={[localStyles.titleContainer, { maxHeight: maxHeight }]}>
                    {editionMode ? (
                        <TitleEdition goal={goal} projectId={projectId} closeTitleEdition={closeTitleEdition} />
                    ) : (
                        <View onLayout={onTitleLayoutChange}>
                            <TitlePresentation
                                projectId={projectId}
                                openTitleEdition={openTitleEdition}
                                goal={goal}
                                disabled={!accessGranted || !loggedUserCanUpdateObject}
                            />
                        </View>
                    )}
                    {showEllipsis && !editionMode && (
                        <Text style={[localStyles.ellipsis, { right: mobile ? 32 : 80 }]}>...</Text>
                    )}
                </View>
            </View>
            {!isFullscreen && (
                <View style={localStyles.bottomHeader}>
                    <TagList
                        completionMilestoneDate={completionMilestoneDate}
                        goal={goal}
                        projectId={projectId}
                        accessGranted={accessGranted}
                        loggedUserCanUpdateObject={loggedUserCanUpdateObject}
                    />
                    <LastEdition projectId={projectId} goal={goal} />
                    <CopyLinkButton style={{ marginRight: 8 }} />
                    <DvBotButton navItem={DV_TAB_GOAL_CHAT} projectId={projectId} assistantId={goal.assistantId} />
                    <OpenInNewWindowButton />
                </View>
            )}
            {isFullscreen && selectedTab === DV_TAB_GOAL_CHAT && (
                <View style={localStyles.bottomHeader}>
                    <BotLine
                        setFullscreen={setFullscreen}
                        objectId={goal.id}
                        assistantId={goal.assistantId}
                        projectId={projectId}
                        objectType={'goals'}
                    />
                </View>
            )}
        </View>
    )
}

const localStyles = StyleSheet.create({
    container: {
        paddingBottom: 24,
    },
    upperHeader: {
        flexDirection: 'row',
        justifyContent: 'space-between',
    },
    titleContainer: {
        marginRight: 'auto',
        flex: 1,
        maxHeight: 800,
        overflowY: 'hidden',
    },
    bottomHeader: {
        marginTop: 22,
        flexDirection: 'row',
        alignItems: 'center',
    },
    backButtonMobile: {
        left: -16,
    },
    ellipsis: {
        ...styles.title4,
        color: colors.Text01,
        backgroundColor: '#ffffff',
        paddingHorizontal: 8,
        position: 'absolute',
        bottom: 0,
        right: 0,
    },
})
