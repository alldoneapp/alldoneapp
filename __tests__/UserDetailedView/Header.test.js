import React from 'react'
import store from '../../redux/store'
import { storeLoggedUserProjects } from '../../redux/actions'
import Header from '../../components/UserDetailedView/Header/Header'

import renderer from 'react-test-renderer'

jest.mock('firebase', () => ({ firestore: {} }));

describe('Header component', () => {
    describe('Header snapshot test', () => {
        it('should render correctly', () => {
            const projects = [{ name: 'Build a Stairway To Heaven', id: '0' }]
            store.dispatch([storeLoggedUserProjects(projects)])
            const tree = renderer.create(<Header projectIndex={0} contact={{ photoURL: 'asd' }} />).toJSON()
            expect(tree).toMatchSnapshot()
        })
    })
})
