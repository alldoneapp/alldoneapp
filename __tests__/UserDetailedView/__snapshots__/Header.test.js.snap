// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Header snapshot test should render correctly 1`] = `
<View
  style={
    Object {
      "flexDirection": "column",
      "height": 140,
      "justifyContent": "space-between",
      "paddingBottom": 24,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "marginRight": "auto",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "height": 64,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "white",
              "height": 32,
            }
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "flex": 1,
              "flexDirection": "row",
              "height": 32,
              "justifyContent": "flex-start",
            }
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 100,
                  "height": 32,
                  "marginLeft": 1,
                  "padding": 3,
                  "width": 32,
                },
                Object {
                  "backgroundColor": undefined,
                },
              ]
            }
          >
            <Image
              source={
                Object {
                  "uri": "asd",
                }
              }
              style={
                Object {
                  "backgroundColor": "#C6CDD2",
                  "borderRadius": 100,
                  "height": 26,
                  "marginRight": 4,
                  "width": 26,
                }
              }
            />
          </View>
          <View
            style={
              Object {
                "marginLeft": 12,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#04142F",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 24,
                    "lineHeight": 32,
                  },
                  Object {
                    "color": "#04142F",
                  },
                ]
              }
            />
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#8A94A6",
                "borderRadius": 16,
                "height": 8,
                "marginHorizontal": 8,
                "width": 8,
              }
            }
          />
          <Text
            style={
              Array [
                Object {
                  "color": "#04142F",
                  "fontFamily": "Roboto-Medium",
                  "fontSize": 24,
                  "lineHeight": 32,
                },
              ]
            }
          >
            Build a Stairway To Heaven
          </Text>
        </View>
      </View>
    </View>
    <View>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": 36,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "Roboto-Medium",
                  "fontSize": 16,
                  "letterSpacing": 0.16,
                  "lineHeight": 24,
                },
                Object {
                  "color": "#8A94A6",
                },
              ]
            }
          >
            PEOPLE
          </Text>
          <View
            style={
              Object {
                "marginLeft": 14,
              }
            }
          >
            <View>
              <Text
                style={
                  Object {
                    "color": "#8A94A6",
                    "fontFamily": "alldone",
                    "fontSize": 20,
                    "opacity": 1,
                  }
                }
              >
                î¨¥
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  />
</View>
`;
